File: qoqo/measurements/cheated_basis_rotation_measurement.py
# Copyright © 2019-2021 HQS Quantum Simulations GmbH. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
# in compliance with the License. You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software distributed under the License
# is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
# or implied. See the License for the specific language governing permissions and limitations under
# the License.
"""Cheated basis rotation measurement"""
from qoqo.measurements._measurement_base_class import MeasurementBaseClass
from qoqo.measurements.measurement_auxiliary_data_input import (
    CheatedBRMeasurementInput
)
from typing import (
    List,
    Optional,
    Union,
    Dict,
    cast
)
from qoqo import Circuit
from qoqo.backends import BackendBaseClass
from qoqo.registers import RegisterOutput
from qoqo.devices import DeviceBaseClass
import numpy as np
import pandas as pd
from hqsbase.qonfig import Qonfig, empty


class CheatedBasisRotationMeasurement(MeasurementBaseClass):
    r"""Cheated basis rotation measurement

    Cheated measurement of expectation values that uses a basis rotation measurement preparation.

    """

    _qonfig_defaults_dict = {

        'backend': {'doc': 'qoqo backend',
                    'default': empty},
        'circuit_list': {'doc': 'The circuits that are run and measured on the backend',
                         'default': None},
        'constant_circuit': {
            'doc': 'Circuit that is always applied before each circuit in the circuit_list',
            'default': None},
        'measurement_input': {'doc': 'Additional input (matrices) required for measurement',
                              'default': Qonfig[CheatedBRMeasurementInput]},
        'device': {'doc': 'Quantum computer device used in the backend',
                   'default': None},
    }

    def __init__(self,
                 backend: Optional[BackendBaseClass] = None,
                 measurement_input: Optional[CheatedBRMeasurementInput] = None,
                 circuit_list: Optional[List[Circuit]] = None,
                 constant_circuit: Optional[Circuit] = None,
                 device: Optional[Union[
                     Qonfig,
                     DeviceBaseClass]] = None,
                 **kwargs) -> None:
        """Initialize measurement

        Args:
            backend: qoqo backend
            measurement_input: Additional input (matrices) required for measurement
            circuit_list: The circuits that are run and measured on the backend
            constant_circuit: Circuit that is always applied before each circuit in the circuit_list
            device: Quantum computer device used in the backend
            kwargs: Additional keyword arguments

        """
        super().__init__(
            circuit_list=circuit_list,
            constant_circuit=constant_circuit,
            backend=backend,
            device=device,
            **kwargs
        )
        if measurement_input is None:
            self.measurement_input = CheatedBRMeasurementInput()
        else:
            self.measurement_input = measurement_input
        self.kwargs = kwargs

    @classmethod
    def from_qonfig(cls,
                    config: Qonfig['CheatedBasisRotationMeasurement']
                    ) -> 'CheatedBasisRotationMeasurement':
        """Create an Instance from Qonfig

        Args:
            config: Qonfig of class

        Returns:
            CheatedBasisRotationMeasurement
        """
        if isinstance(config['backend'], Qonfig):
            init_backend = config['backend'].to_instance(no_copy=True)
        else:
            init_backend = cast(Optional[BackendBaseClass], config['backend'])
        if isinstance(config['device'], Qonfig):
            init_device = config['device'].to_instance(no_copy=True)
        else:
            init_device = cast(Optional[DeviceBaseClass], config['device'])
        clist: List[Circuit] = list()
        for c in config['circuit_list']:
            clist.append(c.to_instance())
        return cls(backend=init_backend,
                   measurement_input=config['measurement_input'].to_instance(no_copy=True),
                   circuit_list=clist,
                   constant_circuit=config['constant_circuit'].to_instance(no_copy=True),
                   device=init_device,
                   )

    def to_qonfig(self) -> 'Qonfig[CheatedBasisRotationMeasurement]':
        """Create a Qonfig from Instance

        Returns:
            Qonfig[CheatedBasisRotationMeasurement]
        """
        config = Qonfig(self.__class__)
        if self.backend is None:
            config['backend'] = None
        else:
            config['backend'] = self.backend.to_qonfig()
        config._values['measurement_input'] = self.measurement_input.to_qonfig()
        config._values['circuit_list'] = [c.to_qonfig() for c in self.circuit_list]
        config._values['constant_circuit'] = self._constant_circuit.to_qonfig()
        if self.device is None:
            config['device'] = None
        else:
            config['device'] = self.device.to_qonfig()

        return config

    def __call__(self, **kwargs) -> pd.Series:
        """Execute measurement

        Args:
            kwargs: Additional keyword arguments

        Returns:
            pd.Series

        """
        if self.backend is None:
            return pd.Series({}, dtype=complex)
        # Dict for all read-out registers
        output_register_dict: Dict[str, RegisterOutput] = dict()
        # Dict for pauli products calculated from each read out register
        for circuit in self._circuit_list:
            self.backend.circuit = self._constant_circuit + circuit
            tmp_output_register_dict = self.backend.run()
            if tmp_output_register_dict is not None:
                output_register_dict.update(tmp_output_register_dict)

        pp_to_exp_val_matrix = self.measurement_input._pp_to_exp_val_matrix

        pauli_products = np.zeros(self.measurement_input._number_pauli_products)
        for register_name, register in output_register_dict.items():
            if '_pauli_product_' in register_name:
                x = register_name.split('_pauli_product_')
                index = int(x[-1])
                pauli_products[index] = register.register[0][0]

        expectation_values = pp_to_exp_val_matrix @ pauli_products

        names = self.measurement_input._measured_exp_vals
        tmp_dict = {('exp_val_' + name): exp_val for name,
                    exp_val in zip(names, expectation_values)}

        if 'global_phase' in output_register_dict.keys():
            tmp_dict['global_phase'] = output_register_dict['global_phase'].register[0][0]

        if not tmp_dict:
            return_series = pd.Series({}, dtype=complex)
        else:
            return_series = pd.Series(tmp_dict)

        return return_series

File: roqoqo/src/measurements/cheated_basis_rotation_measurement.rs
// Copyright © 2021-2022 HQS Quantum Simulations GmbH. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
// in compliance with the License. You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software distributed under the
// License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
// express or implied. See the License for the specific language governing permissions and
// limitations under the License.

use super::*;
use ndarray::Array1;
#[cfg(feature = "serialize")]
use serde::{Deserialize, Serialize};

/// Collected information for executing a cheated measurement of a PauliZ product.
#[derive(Debug, PartialEq, Clone)]
// #[cfg_attr(feature = "json_schema", derive(schemars::JsonSchema))]
#[cfg_attr(feature = "serialize", derive(Serialize, Deserialize))]
pub struct CheatedPauliZProduct {
    /// Constant Circuit that is executed before each Circuit in circuits.
    pub constant_circuit: Option<Circuit>,
    /// Collection of quantum circuits for the separate basis rotations.
    pub circuits: Vec<Circuit>,
    /// Additional input information required for measurement.
    pub input: CheatedPauliZProductInput,
}

impl Measure for CheatedPauliZProduct {
    /// Returns the constant Circuit that is executed before each Circuit in circuits.
    ///
    /// # Returns
    ///
    /// * `&Option<Circuit` - The constant Circuit (None if not defined).
    fn constant_circuit(&self) -> &Option<Circuit> {
        &self.constant_circuit
    }

    /// Returns iterator over circuits for measurement.
    ///
    /// # Returns
    ///
    /// * `Box<dyn Iterator<Item = &'a Circuit> + 'a>` - The quantum circuits.
    fn circuits<'a>(&'a self) -> Box<dyn Iterator<Item = &'a Circuit> + 'a> {
        Box::new(self.circuits.iter())
    }

    /// Returns clone of Measurement with symbolic parameters replaced.
    ///
    /// # Arguments
    ///
    /// * `substituted_parameters` - The HashMap containing the substitutions to use in the Circuit.
    ///
    /// # Returns
    ///
    /// * `Ok(Self)` -  The Circuits with the parameters substituted.
    /// * `Err(RoqoqoError)` - The subsitution failed.
    fn substitute_parameters(
        &self,
        substituted_parameters: HashMap<String, f64>,
    ) -> Result<Self, RoqoqoError> {
        let mut calculator = qoqo_calculator::Calculator::new();
        for (name, val) in substituted_parameters.iter() {
            calculator.set_variable(name, *val)
        }
        let new_constant_circuit = match &self.constant_circuit {
            None => None,
            Some(c) => Some(c.substitute_parameters(&calculator)?),
        };
        let mut new_circuits = Vec::new();
        for circ in self.circuits.iter() {
            let mut calculator = qoqo_calculator::Calculator::new();
            for (name, val) in substituted_parameters.iter() {
                calculator.set_variable(name, *val)
            }
            new_circuits.push(circ.substitute_parameters(&calculator)?)
        }
        Ok(Self {
            constant_circuit: new_constant_circuit,
            circuits: new_circuits,
            input: self.input.clone(),
        })
    }
}

impl MeasureExpectationValues for CheatedPauliZProduct {
    /// Executes the cheated PauliZ product measurement
    ///
    /// # Arguments
    ///
    /// * `bit_registers` - The classical bit registers as a HashMap with the register name as key
    /// * `float_registers` - The classical float registers as a HashMap with the register name as key
    /// * `complex_registers` - The classical complex registers as a HashMap with the register name as key
    ///
    /// # Returns
    ///
    /// * `Ok(Some(HashMap<String, f64>))` - The measurement has been evaluated successfully. The HashMap contains the measured expectation values
    /// * `Ok(None)` - The measurement did not fail but is incomplete. A new round of measurements is needed
    /// * `Err(RoqoqoError)` - Calculator parsing error.
    ///
    #[allow(unused_variables)]
    fn evaluate(
        &self,
        bit_registers: HashMap<String, BitOutputRegister>,
        float_registers: HashMap<String, FloatOutputRegister>,
        complex_registers: HashMap<String, ComplexOutputRegister>,
    ) -> Result<Option<HashMap<String, f64>>, RoqoqoError> {
        let mut pauli_products: Array1<f64> = Array1::zeros(self.input.pauli_product_keys.len());
        for (register_name, register) in float_registers.iter() {
            if let Some(index) = self.input.pauli_product_keys.get(register_name) {
                pauli_products[*index] = register[0][0];
            } else {
                return Err(RoqoqoError::MissingRegister {
                    name: register_name.clone(),
                });
            }
        }
        // Evaluating expectation values
        let mut results: HashMap<String, f64> = HashMap::new();

        for (name, evaluation) in self.input.measured_exp_vals.iter() {
            results.insert(
                name.clone(),
                match evaluation {
                    PauliProductsToExpVal::Linear(hm) => {
                        let mut value: f64 = 0.0;
                        for (index, coefficient) in hm {
                            value += pauli_products[*index] * coefficient;
                        }
                        value
                    }
                    PauliProductsToExpVal::Symbolic(x) => {
                        let mut calculator = qoqo_calculator::Calculator::new();
                        for (ind, p) in pauli_products.iter().enumerate() {
                            calculator.set_variable(format!("pauli_product_{}", ind).as_str(), *p);
                        }
                        calculator.parse_get(x.clone())?
                    }
                },
            );
        }

        Ok(Some(results))
    }
}

impl crate::operations::SupportedVersion for CheatedPauliZProduct {
    fn minimum_supported_roqoqo_version(&self) -> (u32, u32, u32) {
        let mut current_minimum_version = (1, 0, 0);
        let comparison_version = self.input.minimum_supported_roqoqo_version();
        crate::update_roqoqo_version(&mut current_minimum_version, comparison_version);
        if let Some(circuit) = self.constant_circuit() {
            let comparison_version = circuit.minimum_supported_roqoqo_version();
            crate::update_roqoqo_version(&mut current_minimum_version, comparison_version);
        }
        for circuit in self.circuits.iter() {
            let comparison_version = circuit.minimum_supported_roqoqo_version();
            crate::update_roqoqo_version(&mut current_minimum_version, comparison_version);
        }
        current_minimum_version
    }
}
